### ISSUE TITLE
Auto configure BIND9 for no DNS downtime for October DNSSEC KSK rollover of root zone

### ISSUE THREAD

**Issue by chris001** [webmin/webmin#617]
ICANN is finally rolling the Root Zone's DNSSEC Key Signing Key in October 2017.  The KSK currently in use has not been rolled since it was first published in 2010.

Unless all Webmin's BIND9 installs are upgraded to a version supporting RFC5011 AND configured to automatically roll the key when the old key is unpublished in October, then all DNS resolution will stop working causing major outages.

It's a relatively simple fix to get it working right now so that no downtime happens later.

How to config BIND9 to support RFC5011 automated trust anchor update protocol for no DNS downtime during the upcoming DNSSEC Key Signing Key rollover for the root zone?

**Executive summary**

If you manage a DNS resolver, you may need to take action in 2017 due to the upcoming root key rollover.

* If you use BIND with `managed-keys` for the root zone or `dnssec-validation auto`, there is low risk.
* If you use BIND with `trusted-keys` for the root zone, you need to update your configuration.
* Anyone setting up a new BIND instance around the time of the root key rollover will need to pay careful attention, to ensure their system is able to initialize properly.
* Organizations repackaging or redistributing BIND will need to update their distributions in 2017 to ensure any new installations that happen during or after October 2017 include the new key.

There are some risks in all configurations, discussed below:
https://www.icann.org/resources/pages/ksk-rollover

* https://www.isc.org/blogs/2017-root-key-rollover-what-does-it-mean-for-bind-users/
* https://kb.isc.org/article/AA-01525/30/Root-KSK-Rollover-in-BIND.html
* ICANN official testbed for the KSK rollover https://automated-ksk-test.research.icann.org/

**Comment by jcameron** [comment]
So is there something Webmin needs to do beyond offering the latest BIND package update?

**Comment by chris001** [comment]
Ya, as far as I can see, 
1. Webmin should modify settings of BIND9 so that the following conditions are met so as to avoid catastrophic halt of DNS name resolution:
* if webmin/virtualmin is running bind9 in `trusted-keys` mode for the root zone then BIND9 will stop resolving DNS names when that old DNSSEC key rolls out of the ICANN root zone ( `.` ) leaving only the new key.  
* Also, there's a risk of DNS not resolving when BIND9 is not running in `managed-keys` for the root zone or nor `dnssec-validation auto`.
Basically, this is kind of a big deal, ICANN people spent 18 months preparing for every possible scenario and for this key roll of the DNSSEC Key Signing Key of the internet's root zone!

2. Also, consider urging or automatically forcing an upgrade to the newest version of BIND9.  Relatively certain but this needs more research to verify completely: Some Webmin systems could be running old versions of BIND9 which do not even have this automatic key rolling ability.  The automatic management of DNSSEC keys was added to BIND9 recently. So some Webmin systems may experience an outage of DNS even if Webmin modifies these settings in the `bind` config to try to make bind9 auto switch over to the new key.

* From the BIND9 9.11.1-P2 release notes:
https://deepthought.isc.org/article/AA-01507/0/BIND-9.11.1-P2-Release-Notes.html
**New DNSSEC Root Key**

ICANN is in the process of introducing a new Key Signing Key (KSK) for the global root zone. BIND has multiple methods for managing DNSSEC trust anchors, with somewhat different behaviors. If the root key is configured using the `managed-keys` statement, or if the pre-configured root key is enabled by using `dnssec-validation auto`, then BIND can keep keys up to date automatically. Servers configured in this way will roll seamlessly to the new key when it is published in the root zone. However, keys configured using the `trusted-keys` statement are not automatically maintained. If your server is performing DNSSEC validation and is configured using `trusted-keys`, you are advised to change your configuration before the root zone begins signing with the new KSK. This is currently scheduled for October 11, 2017.

This release includes an updated version of the `bind.keys` file containing the new root key. This file can also be downloaded from https://www.isc.org/bind-keys .

**Comment by chris001** [comment]
Webmin is using the older `trusted-keys` to specify the root zone key as a static literal value, so all DNS lookup will definitely start to fail in October 2017 unless changed before then.

[![DNSSEC new key signing ceremony](https://img.youtube.com/vi/b9j-sfP9GUU/0.jpg)](https://www.youtube.com/watch?v=b9j-sfP9GUU)

https://www.isc.org/blogs/2017-root-key-rollover-what-does-it-mean-for-bind-users/

In Summary ...
* If you are running a BIND validating resolver using `trusted-keys` and you can upgrade to `managed-keys`, do so now, before the root key rollover
* https://www.icann.org/resources/pages/ksk-rollover/#resources
* Be on the lookout for possible DNSSEC validation failures on or around, October 11th 2017, in case of unforeseen problems.

Anyone providing re-packaged versions of BIND should plan to update their distributions in Q1 or Q2 of 2017.
Older versions of BIND, for example in older operating system releases, must also be updated.

**What to do if you are using BIND with trusted-keys**

If you can, the best thing would be to update your configuration to use `managed-keys`, at least for the root zone key.  If you choose this option, it is essential to make this change before the key rollover occurs in October, so that BIND can successfully initialize managed-keys with the 2010 key prior to the rollover.

The easiest way to enable `managed-keys` is to simply use the shortcut, `dnssec-validation auto;`   You must also have a `bind.keys` file with the 2010 key included, you can get this by downloading a fresh copy of BIND, or by just downloading and installing the `keys` file itself.  To initialize the `managed-keys` database, restart BIND with `managed-keys` enabled and a valid `bind.keys` file present. _**If you do this prior to the October 2017 rollover, BIND will validate the 2010 key and self-maintain during the rollover.**_  

If you want to continue using `trusted-keys` for the root zone, you must ensure your `bind.keys` file is updated to add the 2017 root key before the rollover occurs.  You can’t do anything to prepare until the new key is published. To get the new `bind.keys` file after the 2017 key is published:

1. If you download and run an updated version of BIND from ISC* after the new key is published (currently planned for February, 2017), the new key will be included. (* while we will update the software published by ISC, and we will alert the operating system packagers to update their packages, we have no control over what is provided in the various operating system packages)   If you then restart with a `managed-keys` configuration (either using `managed-keys` or `dnssec-validation auto`) BIND will initialize with the included key and self-maintain during the key rollover.
2. If you don’t want to update your BIND software, you can download just the new key from ISC or ICANN and manually add it to the `bind.keys` file in `/etc` (it is moved there after installation), or replace the `bind.keys` file with a new one downloaded from `isc.org`, and restart BIND. This will cause BIND to load the keys in the new `bind.keys` file.  Be sure to update the `bind.keys` file in your ‘master copy’ of BIND if you have a provisioning system creating builds for new installations.
3. Keep the 2010 key in the trust anchor database during the overlap and rollover period. When the old key is withdrawn, you may remove the old key (you don’t have to and it might be wisest to leave it in there just in case).

**Comment by jcameron** [comment]
This sounds pretty serious ... although some of those actions seem outside the scope of Webmin, or are already supported (like offering to install a new BIND package if available).

Is the way Webmin currently sets of DNSSEC wrong though?

**Comment by chris001** [comment]
Yes what webmin does wrong regarding DNSSEC is webmin adds the setting `trusted-keys` to `/etc/bind/named.conf`.  This will absolutely break during the root zone KSK key rollover.

From now on, all new installs of webmin should use `dnssec-validation auto;' in place of `trusted-keys`, and all existing installs of webmin should probably be changed by the webmin upgrader script, to comment out or remove the setting `trusted-keys` and instead insert the setting `dnssec-validation auto;`, then verify that the `bind.keys` file from the package is present, and restart bind.

From above:
**What to do if you are using BIND with trusted-keys**

If you can, the best thing would be to update your configuration to use `managed-keys`, at least for the root zone key. If you choose this option, it is essential to make this change before the key rollover occurs in October, so that BIND can successfully initialize `managed-keys` with the 2010 key prior to the rollover.

The easiest way to enable `managed-keys` is to simply use the shortcut, `dnssec-validation auto;` You must also have a `bind.keys` file with the 2010 key included, you can get this by downloading a fresh copy of BIND, or by just downloading and installing the `bind.keys` file itself. To initialize the `managed-keys` database, restart BIND with `managed-keys` enabled and a valid `bind.keys` file present. If you do this prior to the October 2017 rollover, BIND will validate the 2010 key and self-maintain during the rollover.

**Comment by jcameron** [comment]
Ok, thanks - I'll do that.

Webmin also has support for setting the `dnssec-lookaside` and additional `trusted-keys` directives. Are these completely obsolete in the modern DNSSEC world?

**Comment by chris001** [comment]
"Webmin also has support for setting the `dnssec-lookaside` and additional `trusted-keys` directives. Are these completely obsolete in the modern DNSSEC world?"

From here:
https://www.isc.org/downloads/bind/bind-keys/

For Current Releases (BIND 9.8 and higher):

*    If you configure your own `managed-keys` statement in `named.conf`, this will take precedence.
*    If you put `dnssec-validation auto` in `named.conf`, `named` will read the root key from `bind.keys` the first time it executes.
*    If you put `dnssec-lookaside auto` in `named.conf`, `named` will read the DLV key from `bind.keys` the first time it executes.
*    If you don’t have anything in `named.conf` and there is no `bind.keys` file, `named` will use the compiled in keys.

Note: these are managed keys, so this is only applies the first time you execute `named`. Assuming that the keys are not already expired (in which case named will log that the key is expired and validation will not work), named will use RFC 5011 to detect new keys and automatically roll and maintain keys. Once `named` is managing the keys, the current keys will be in `managed-keys.bind` or `*.mkeys`, if you use views.

(From the previous paragraph) "(note that the DLV has been decommissioned and we recommend updating resolver configurations that query the DLV)"

So it appears this `dnssec-lookaside` directive is possibly no longer used or needed, because the documentation on this page says the DLV has been decommissioned.  RFC 5011 is able to automatically handle key rolls from now onward into the future, as long as the first key at startup/restart, used on bind's first run right now (key located in `bind.keys`) is matching a valid key in the root zone, and the directive `dnssec-validation auto` is used.  As of right now, both 2010 and 2017 keys are in the root zone.  A few months from now - October or January - the root zone will only have the 2017 key.

**Comment by jcameron** [comment]
Maybe I should completely remove support for editing the `trusted-keys` directive then, and perhaps also `dnssec-lookaside` ?

**Comment by chris001** [comment]
Good idea, just to save people from shooting themselves in the foot and misconfiguring bind by accident and breaking validated security on their bind recursive resolver. They could always edit the file manually if they really wanted to get those old fashioned settings in there, although I can't imagine why anyone would want to, maybe researchers would, to see if they could break something.

**Comment by marka63** [comment]
Removing DLV will stop some of the thousands of useless queries to dlv.isc.org that ISC receives every day.   The zone is empty and it has been for years.  It's only still there to absorb the traffic from sites running on autopilot.

**Comment by iliajie** [comment]
Jamie, are we still adding the `dnssec-lookaside` record?

**Comment by chris001** [comment]
Maybe the next webmin installer/updater could actively disable or remove DNS Lookaside Validation globally in BIND9, which will stop the pointless lookup traffic to `dlv.isc.org`.  This article shows precisely how:
https://kb.isc.org/docs/disable-dnssec-lookaside-dlv-now-heres-how

**Comment by jcameron** [comment]
Pretty sure we are no longer adding this record

**Comment by iliajie** [comment]
But I still see occurrences of it in the BIND module, particularly in `fix_trusted.cgi` and `save_trusted.cgi` files under the `Save DLV zones` comment.

**Comment by jcameron** [comment]
Ah yes ... I've checked in a fix to remove those.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
